### PR TITLE
Only replace on changed node selector if pod spec hash has changed

### PR DESCRIPTION
**Pull request by benzim**
# Description

This is a fix for https://github.com/FoundationDB/fdb-kubernetes-operator/issues/950.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Discussion

This will trigger a replacement if any part of the pod template spec is changed if there is a [PodNodeSelector](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podnodeselector). Which isn't ideal but is better than the previous behavior of the pods being replaced endlessly.

# Testing

I ran unit tests and manually deployed the operator and cluster in a namespace with a PodNodeSelector.

### CODE REVIEW

File: internal/replacements/replacements.go [hunk 1]
```diff
@@ -189,13 +189,8 @@ func instanceNeedsRemoval(cluster *fdbtypes.FoundationDBCluster, pod *corev1.Pod
...
 
 	expectedNodeSelector := cluster.GetProcessSettings(processClass).PodTemplate.Spec.NodeSelector
-	if !equality.Semantic.DeepEqual(pod.Spec.NodeSelector, expectedNodeSelector) {
-		logger.Info("Replace instance",
-			"reason", fmt.Sprintf("nodeSelector has changed from %s to %s", pod.Spec.NodeSelector, expectedNodeSelector))
-		return true, nil
-	}
 
-	if cluster.Spec.UpdatePodsByReplacement {
+	if cluster.Spec.UpdatePodsByReplacement || !equality.Semantic.DeepEqual(pod.Spec.NodeSelector, expectedNodeSelector) {
```
Comment (johscheuer) [inline internal/replacements/replacements.go:193]: Can we still leave the node selector check in a different if statement? That makes it easier to see why the Pod was replaced with this change we will only see "specHash has changed from ... to ..." but we can be more precise by logging that the node selector doesn't match.

**Review by johscheuer — CHANGES_REQUESTED** [review]

**Review by johscheuer — COMMENTED** [review]
The changes look good to me, could you add an additional test case here: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/internal/replacements/replacements_test.go#L238-L255 where we change the nodeSelector of the Pod. That shouldn't result in a replacement with the new changes.

**Comment by benzim** [comment]
@johscheuer Hopefully this isn't out of scope for this PR, but the replacements tests weren't running when I ran `make test` so I added some stuff to get the replacements tests running and passing.

**Review by johscheuer — APPROVED** [review]
LGTM 👍 Thanks for fixing the test suite.

**Pull request merged**